Updates for GUN 0.8 API changes

sjones6 opened the pull request:
@PsychoLlama : here's a start. `HAM.union` will need replaced in `Adapters/index.js`. More tomorrow.

amark commented:
@sjones6 good observation, yes, I indeed changed HAM.union, I'll see if I can find a good enough replacement.

amark commented:
This code should work to replace `Gun.HAM.union`:

```
function union(vertex, node, opt){
	if(!node || !node._){ return }
	vertex = vertex || Gun.state.to(node);
	if(!vertex || !vertex._){ return }
	opt = Gun.num.is(opt)? {machine: opt} : {machine: Gun.state()};
	opt.union = Gun.obj.copy(vertex); // Slow performance.
	if(!Gun.node.is(node, function(val, key){
		var HAM = Gun.HAM(opt.machine, Gun.state.is(node, key), Gun.state.is(vertex, key, true), val, vertex[key]);
		if(!HAM.incoming){ return }
		Gun.state.to(node, key, opt.union);
	})){ return }
	return opt.union;
}
```

All the methods used inside of their should also probably be stable now too, I think/hope.

(except jokingly: maybe I should rename the HAM return values :P to something sexier than `incoming`. This would literally break everything everywhere. But but but... prettier names would be so niiiiiiice.)

I will consider adding this function somewhere to GUN core maybe.

sjones6 commented:
@amark : Thanks! That did the trick :) You know what they say about naming in programming ...

@PsychoLlama : I've updated the tests with the latest API changes (i.e., removal of `path` and `not`; all now passing. Updated README as well. Locally, gun-level is working with mongodown and read/writes are working as expected.

Whenever you merge in and deploy, probably need to bump up a major version? Theoretically, this should be compatible with previous versions (<0.8) but I didn't test to confirm. If not, that probably warrants a note in the README or something.

PsychoLlama commented on package.json:
```diff
@@ -23,8 +23,7 @@
   "license": "MIT",
   "dependencies": {
-    "babel-runtime": "^6.11.6",
-    "gun": "^0.7.0"
```
Looks like this broke [the build](https://travis-ci.org/PsychoLlama/gun-level/jobs/252397528), but it should be an easy workaround. If you add gun as a `devDependency` the tests will pass, and developers installing the package won't get it automatically. I'd recommend putting the same entry in `peerDependencies`, that way people get a warning if they're using an incompatible gun version, but it won't be bundled with the package.

sjones6 replied:
@PsychoLlama : Cool. I've added that back in to `devDependencies` and `peerDependencies`.

PsychoLlama commented on src/Adapter/index.js:
```diff
@@ -6,7 +6,19 @@ const options = {
+		Gun.state.to(node, key, opt.union);
+	})){ return }
+	return opt.union;
+}
```
I suspect this is why [eslint freaked out](https://travis-ci.org/PsychoLlama/gun-level/jobs/252397528). I'm glad you added this into gun-level, now future breaking changes in gun core shouldn't affect this.

I would not object if you put `/* eslint-disable */` before the union code block and ended it with `/* eslint-enable */`. That should fix the rest of the build.

sjones6 replied:
@PsychoLlama : I've added those lines to disable linting for that block. I'm usually militant about linting stuff, so I may come back and fix that up if time permits in the future. :)

PsychoLlama replied:
I feel ya. Most of my builds fail if the lint doesn't pass. A drawback I've seen is it adds another barrier to entry for contributors. This might be a good project to experiment with [prettier](https://github.com/prettier/prettier) on 🤔

sjones6 replied:
Cool. I've also wanted to give Prettier a shot. It'd be sweet as a build step so contributors don't even really have to think about it.

PsychoLlama commented on README.md:
```diff
@@ -87,28 +88,41 @@ const bob = gun.get('bob').put({ name: 'Bob' })
 
-const levelDB = levelup('localhost', {
+// Initialize Level
+const levelDB = levelup('localhost:27017/YOUR_COLLECTION_NAME', {
```
Thanks for fixing this!!!

sjones6 replied:
Yep!

sjones6 commented:
@PsychoLlama : Thanks for the review! Fixes are pushed. We'll see if the build passes now.

sjones6 commented:
@PsychoLlama : I just re-checked some things here, and it seems `Gun.on('opt', callback)` isn't firing when it's in the devDependencies either... not sure why. I removed and it worked again.

Try dropping a debug statement or breakpoint in there and see if you can confirm.

I'll be back on later to re-check.

PsychoLlama commented:
@sjones6 this looks great! I pulled your fork and did a manual test to see if I could replicate the `Gun.on('opt')` issue, but everything seems to be working fine 🤷‍♂️  (saving data, closing the process and recovering from level).

I'm up for merging and releasing these changes as v6. Any objection?

sjones6 commented:
@PsychoLlama : If you've tested and it's working, I say merge it in.

It's probably something on my machine. I'll keep digging and see if I find the issue. Others can open issues if they run into problems I suppose.

PsychoLlama commented:
Cool, let's merge this thing!

Here's some common things I usually miss, though they might not apply for you:
- Run the build again, just to be sure
- `rm -rf` the node modules and reinstall with `yarn --pure-lockfile`
- plead to a deity

Maybe one of those will fix the thing
